### ISSUE TITLE
Fix hang inside Docker

### ISSUE THREAD

**Issue by mikestreety** [stylelint/stylelint#7046]
### What minimal example or steps are needed to reproduce the bug?

For reasons that are beyond the scope of this issue, I am looking to Docker-ise Stylelint and some configuration. It seems running it in inside Docker hangs when you point it to a valid docker path.

**Dockerfile**

```dockerfile
FROM node:20-alpine3.17

RUN npm install -g stylelint stylelint-config-standard
````

Then running & accessing the docker container:

```bash
docker run -it --rm -v $(pwd):/app scss-stylelint-dev "/bin/sh"
```

Then running stylelint inside the container

```bash
stylelint app/**/*.css --config-basedir /usr/local/lib/node_modules/
```

Running this on my computer is near-instant, but inside the docker container this hangs for a long time - sometimes it is successful, other times not.

I am 100% happy to blame this on my dockering skills, but I've tried everything I can think of (yarn, instakling in different places etc)

### What minimal configuration is needed to reproduce the bug?

```json
{
  "extends": "/usr/local/lib/node_modules/stylelint-config-standard"
}
```

### How did you run Stylelint?

```shell
stylelint app/**/*.css --config-basedir /usr/local/lib/node_modules/
```

### Which Stylelint-related dependencies are you using?

- stylelint@^13.0.0
- stylelint-config-standard@^34.0.0

### What did you expect to happen?

It to lint the CSS

### What actually happened?

It hangs when given a valid path. If the glob path is invalid, it exists quickly

**Comment by jeddy3** [comment]
@mikestreety Thanks for the report and for using the template.

Try quoting the glob:

```console
stylelint "app/**/*.css" --config-basedir /usr/local/lib/node_modules/
```

So that Stylelint processes it rather than the terminal running inside Docker. That's the only thing I can think of offhand.

**Comment by mikestreety** [comment]
Thanks for the quick response @jeddy3, unfortunately that hasn't made any difference in my test.

Edit: Sorry, this turned into a bit of a long comment

Is there some debugging I can enable to see what it is hanging on? It might be finding the files quickly but may be getting stuck with something else in the process. 

<details>
<summary>Timings of running commands</summary>

I'm just thinking out loud here and thought I would get some proper timings for it:

**With a Glob**

```
$ time stylelint "assets/**/*.css" --config-basedir /usr/local/lib/node_modules/

real    0m 37.05s
user    0m 37.84s
sys     0m 0.75s
```

**With a specific file (78.5Kb)** - this did have a lot of warnings

```
$ time stylelint "html/assets/css/screen.css" --config-basedir /usr/local/lib/node_modules/

real    0m 37.53s
user    0m 38.04s
sys     0m 0.88s
````

Seems the glob has no difference...

**With a smaller specific file (1.7Kb)**

```
$ time stylelint "html/assets/css/print.css" --config-basedir /usr/local/lib/node_modules/

real    0m 1.11s
user    0m 1.09s
sys     0m 0.26s
```

---

These tests are a little unfair as they are minified CSS files (I use stylescss lint), so there are plenty of warnings generated. I thought I would use a [CSS sheet from a public repo](https://github.com/contentful/11ty-contentful-starter/blob/master/assets/css/main.css) to see what happens

```
$ time stylelint "test.css" --config-basedir /usr/local/lib/node_modules/

real    0m 7.90s
user    0m 8.79s
sys     0m 0.31s
```

This is a bit more manageable but is still longer than I would expect.

To give some comparisons:

**Running stylelint locally**

```
$ time stylelint **/*.scss --color --config /config/Stylelint.json --config-basedir /usr/local/lib/node_modules/  --ignore-path /config/Stylelint-Ignore

1.13s user
```

**Running stylelint in Docker** (on the same directory)** (I cancelled it...)

```
time stylelint app/**/*.scss --color --config /config/Stylelint.json --config-basedir /usr/local/lib/node_modules/  --ignore-path /config/Stylelint-Ignore
^C
real    1m 26.04s
user    1m 26.70s
sys     0m 0.67s
```

</details>

**Comment by jeddy3** [comment]
Thanks for diving into it some more, particularly the timings. I'm unfamiliar with Docker, but someone else may have an idea. This is the first report of a performance issue inside Docker. I wonder if that's because not many people Dockerise Stylelint or because their configurations were fast.

**Comment by romainmenke** [comment]
@mikestreety Can you share anything about you local system?
Is this linux, mac, ... What is the filesystem of your local drive?

Does this config and how you run it apply autofixes? i.e. are there any local filesystem writes?

**Comment by mikestreety** [comment]
@romainmenke 

- My local system is Mac (Ventura 13.4)
- No fixes are applied unless `--fix` is appended

**Comment by romainmenke** [comment]
Is it an APFS volume? (most moderns macs with default setups will have this)
You can check this in disk utility.

<img width="326" alt="Screenshot 2023-07-05 at 12 40 43" src="https://github.com/stylelint/stylelint/assets/11521496/1bac4eae-5ffa-4b77-bd3e-c4973ed2fc89">

**Comment by mikestreety** [comment]
Looks like it: 
<img width="358" alt="image" src="https://github.com/stylelint/stylelint/assets/354085/3b2b451f-b25f-4a7a-b92f-1e1902124627">

**Comment by romainmenke** [comment]
This is caused by the formatter.

<img width="1177" alt="Screenshot 2023-07-05 at 13 27 20" src="https://github.com/stylelint/stylelint/assets/11521496/ecd8b8bd-c5b2-4c5e-b35e-bff74c33d671">

I suspect this tries to determine the width of the output terminal?
I can imagine that this extremely slow for Dockerized workloads.

Setting a different formatter should make it fast again : `--fo
rmatter unix`

**Comment by mikestreety** [comment]
No luck, unfortunately. I left it for nearly 7 minutes.

One thing  I have got is a slightly outdated version - I'm running 13 (latest is 15) and `stylelint-scss` is 3 (latest 5).

I'll try upgrading and see if that makes any difference

**Comment by romainmenke** [comment]
To profile you could reduce the size of the CSS that is linted and verify that it can finish.

Then run :

- `cd /app`
- `node --cpu-prof /usr/local/bin/stylelint ./**/*.css`

The cpu profile will then appear locally so that you can inspect it.
VSCode has builtin support for these files.

That will tell you which thing is slow.

-----

I am making the assumption that it never truly hangs, but that it is just taking a very long time. I suspect this to be the case because I would expect hangs to be consistent, not intermittent.

I might be wrong about this.

**Comment by mikestreety** [comment]
You're correct @romainmenke In that reducing the number of files for it to search for does speed it up.

I removed most of the files (leaving just 8) and it took over 4 seconds to process. Hopefully you can make a bit more sense out of this graph than I can, but that "Script" drills down all the way to 

```
> normalizeAllRuleSettings /usr/local/lib/node_modules/stylelint/lib/normalizeAllRuleSettings.js:14
```

![image](https://github.com/stylelint/stylelint/assets/354085/53923e16-bc00-4eb1-b55e-db59c57e7a67)

**Comment by romainmenke** [comment]
Can you try with a few more files?
Ideally something that takes +15 seconds.

Could you then drop the cpu profile in a gist?
Then I can inspect it in detail :)

**Comment by jeddy3** [comment]
> One thing I have got is a slightly outdated version - I'm running 13 (latest is 15) and stylelint-scss is 3 (latest 5).
> I'll try upgrading and see if that makes any difference

Yes, please upgrade. The last `13.x` release was [over 2 years ago](https://github.com/stylelint/stylelint/releases/tag/13.13.1). There have been [many releases since then](https://github.com/stylelint/stylelint/releases).

**Comment by romainmenke** [comment]
There is already a second of work being done with only loading Stylelint and its dependencies. This is a lot longer than it should take and a good indication that the problem is not with Stylelint.

<img width="216" alt="Screenshot 2023-07-05 at 18 02 40" src="https://github.com/stylelint/stylelint/assets/11521496/70a65d7d-6eef-4d34-9e96-42653c37fcb5">

The same work on my machine only takes ±150ms inside Docker :

<img width="224" alt="Screenshot 2023-07-05 at 18 06 27" src="https://github.com/stylelint/stylelint/assets/11521496/6c821be3-95f9-4f52-b161-d85fe1dd58fc">

This could be the encrypted disk, or configuration in Docker.

**Comment by jeddy3** [comment]
> Upgrading stylelint has dropped it down to 6 seconds

It's great to know that all our performance work in `14.x` and `15.x` has born fruit.

> There is already a second of work being done with only loading Stylelint and its dependencies. This is a lot longer than it should take and a good indication that the problem is not with Stylelint.

I think we can close the issue then.

@mikestreety If you do find the root cause, you're welcome to share it here as it may benefit someone else.

**Comment by mikestreety** [comment]
Thanks again for all your assistance @jeddy3 and @romainmenke. It was hugely appreciated and although, I was wasting your time you made me feel like I wasn't 😄 

This is all for CI purposes and in order to push it, it needed to be made with a slightly different platform (Mac M1 going to Debian Intel). Because of this, I was building and running the image with the `--platform` attributes. Removing this and building a slightly more optimised docker image has got this down further (and looks similar to your experience)

So **ultimately** everything was _my_ fault. Sorry!

**TL:DR;**

- Make sure you are up-to-date with (at time of writing) Version 15
- Build the docker image for your platform
- If you wish to debug, put `node --cpu-prof` in front of your command

Thanks again 😃
